File: QoSI/TP4/legion/Go/src/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"os/signal"
	"time"

	"github.com/bwmarrin/discordgo"
	"gopkg.in/yaml.v2"
)

// Config holds all configuration for our Discord bot
type Config struct {
	Channels    map[string]string `yaml:"channels"`
	Credentials struct {
		Discord string `yaml:"discord"`
	} `yaml:"credentials"`
	Guild struct {
		ID string `yaml:"id"`
	} `yaml:"guild"`
}

// NewConfig reads configuration from a YAML file and returns a configuration object and an error
func NewConfig(filename string) (*Config, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	var config Config
	err = yaml.Unmarshal(data, &config)
	if err != nil {
		return nil, err
	}

	return &config, nil
}

func main() {
	config, err := NewConfig("config.yaml")
	if err != nil {
		fmt.Printf("Failed to load configuration: %v\n", err)
		return
	}

	dg, err := discordgo.New("Bot " + config.Credentials.Discord)
	if err != nil {
		fmt.Printf("Error creating Discord session: %v\n", err)
		return
	}

	dg.AddHandler(func(s *discordgo.Session, r *discordgo.Ready) {
		fmt.Println("Bot is ready. Setting up handlers...")
		go readyHandler(s, config)
	})

	err = dg.Open()
	if err != nil {
		fmt.Printf("Error opening connection to Discord: %v\n", err)
		return
	}
	defer dg.Close()

	fmt.Println("Bot is now running. Press CTRL-C to exit.")

	stop := make(chan os.Signal, 1)
	signal.Notify(stop, os.Interrupt)
	<-stop

	fmt.Println("Bot is shutting down.")
}

func readyHandler(s *discordgo.Session, config *Config) {
	publicIP := getPublicIP()
	fmt.Printf("Resolved public IP: %s\n", publicIP)

	hostname := "zephyrus"
	fmt.Printf("Resolved hostname: %s\n", hostname)

	latency := s.HeartbeatLatency().Seconds() * 1000
	fmt.Printf("Resolved latency: %fms\n", latency)

	timestamp := time.Now().Unix()
	jsonMessage := fmt.Sprintf("{\"latency\": %f, \"public_ip\": \"%s\", \"hostname\": \"%s\", \"lang\": \"go\", \"timestamp\": %d}", latency, publicIP, hostname, timestamp)
	fmt.Println("Formatted JSON message:", jsonMessage)

	channelID, exists := config.Channels[hostname]
	if exists {
		fmt.Printf("Attempting to send a message to channel ID %s\n", channelID)
		_, err := s.ChannelMessageSend(channelID, fmt.Sprintf("```json\n%s\n```", jsonMessage))
		if err != nil {
			fmt.Printf("Failed to send message to channel %s: %v\n", channelID, err)
		} else {
			fmt.Println("Message sent successfully.")
		}
	} else {
		fmt.Printf("No channel configured for hostname: %s\n", hostname)
	}

	// Command to close the session and exit the program
	s.Close()
	os.Exit(0) // Exit after sending the message
}

func getPublicIP() string {
	resp, err := http.Get("https://api.ipify.org")
	if err != nil {
		fmt.Printf("Failed to get public IP: %v\n", err)
		return "Error"
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("Failed to read response body: %v\n", err)
		return "Error"
	}
	return string(body)
}

File: QoSI/TP4/legion/Rust/src/main.rs
use serenity::{
    async_trait,
    framework::standard::{
        StandardFramework,
        CommandResult,
        macros::{command, group},
        Args,
    },
    client::{bridge::gateway::ShardManager, Client, Context},
    model::{channel::Message, gateway::Ready, id::ChannelId},
    prelude::*,
};
use reqwest;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::time::sleep;

#[derive(Debug, Deserialize, Serialize)]
struct Config {
    channels: HashMap<String, String>,
    credentials: Credentials,
    guild: Guild,
}

#[derive(Debug, Deserialize, Serialize)]
struct Credentials {
    discord: String,
}

#[derive(Debug, Deserialize, Serialize)]
struct Guild {
    id: String,
}

struct Handler {
    config: Config,
}

impl TypeMapKey for MyShardManagerContainer {
    type Value = Arc<Mutex<MyShardManager>>;
}

struct MyShardManagerContainer;

struct MyShardManager {
    shard_manager: Arc<Mutex<ShardManager>>,
    runners: HashMap<u64, Runner>,
}

struct Runner {
    id: u64,
    latency: Option<Duration>,
}

#[group]
#[commands(ping, shutdown)]
struct General;

#[command]
async fn ping(ctx: &Context, msg: &Message) -> CommandResult {
    msg.reply(ctx, "Pong!").await?;
    Ok(())
}

#[command]
async fn shutdown(ctx: &Context, msg: &Message, _: Args) -> CommandResult {
    if let Ok(current_user) = ctx.http.get_current_user().await {
        if msg.author.id == current_user.id {
            let data_read = ctx.data.read().await;
            if let Some(shard_manager) = data_read.get::<MyShardManagerContainer>() {
                let mut manager = shard_manager.lock().await;
                manager.shard_manager.lock().await.shutdown_all().await;
                msg.reply(ctx, "Shutting down...").await?;

                // Signal that the bot should shut down.
                return Ok(());
            }
            msg.reply(ctx, "Failed to fetch shard manager.").await?;
        } else {
            msg.reply(ctx, "You do not have permission to perform this operation.").await?;
        }
    } else {
        msg.reply(ctx, "Failed to fetch current user information.").await?;
    }
    Ok(())
}

#[async_trait]
impl EventHandler for Handler {
    async fn ready(&self, ctx: Context, _ready: Ready) {
        println!("Bot is ready. Setting up handlers...");
        sleep(Duration::from_secs(5)).await;

        let public_ip = match reqwest::get("https://api.ipify.org").await {
            Ok(resp) => resp.text().await.unwrap_or_else(|_| "Error fetching IP".to_string()),
            Err(_) => "Error fetching IP".to_string(),
        };

        if let Some(shard_manager_container) = ctx.data.read().await.get::<MyShardManagerContainer>() {
            let mut shard_manager = shard_manager_container.lock().await;
            shard_manager.runners.insert(0, Runner { id: 0, latency: Some(Duration::from_millis(150)) });
            println!("Simulated latency set for Runner ID 0");
        }

        let latency = calculate_latency(&ctx).await;
        println!("Calculated latency: {:.2}ms", latency);

        let hostname_str = "zephyrus";
        println!("Using static hostname: {}", hostname_str);

        let timestamp = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis();
        let json_string = format!(
            "```json\n{{\"latency\": {:.2}, \"public_ip\": \"{}\", \"hostname\": \"{}\", \"lang\": \"rust\", \"timestamp\": {} }}\n```",
            latency,
            public_ip,
            hostname_str,
            timestamp
        );

        if let Some(channel_id_str) = self.config.channels.get(hostname_str) {
            if let Ok(channel_id) = channel_id_str.parse::<u64>() {
                let channel_id = ChannelId(channel_id);
                if let Err(why) = channel_id.say(&ctx.http, &json_string).await {
                    println!("Error sending message: {:?}", why);
                } else {
                    println!("Message sent successfully to channel ID {}", channel_id);
                    // Desligar imediatamente após enviar a mensagem
                    let data_read = ctx.data.read().await;
                    if let Some(shard_manager) = data_read.get::<MyShardManagerContainer>() {
                        let mut manager = shard_manager.lock().await;
                        manager.shard_manager.lock().await.shutdown_all().await;
                    }
                    return; // Encerra a função ready para não continuar o processamento
                }
            } else {
                println!("Invalid channel ID format in config");
            }
        } else {
            println!("No channel configured for hostname: {}", hostname_str);
        }
    }
}

async fn calculate_latency(ctx: &Context) -> f64 {
    let data_read = ctx.data.read().await;
    if let Some(shard_manager_container) = data_read.get::<MyShardManagerContainer>() {
        let shard_manager = shard_manager_container.lock().await;

        let latencies: Vec<f64> = shard_manager.runners.values()
            .filter_map(|runner| runner.latency)
            .map(|latency| latency.as_secs_f64() * 1000.0)
            .collect();

        let total_latency: f64 = latencies.iter().sum();
        let count = latencies.len() as f64;

        if count > 0.0 {
            return total_latency / count;
        }
    }
    0.0  // Default return value
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config: Config = serde_yaml::from_reader(std::fs::File::open("config.yaml")?)?;

    let framework = StandardFramework::new()
        .configure(|c| c.prefix("!"))
        .group(&GENERAL_GROUP);

    let handler = Handler { config };
    let intents = GatewayIntents::GUILDS | GatewayIntents::GUILD_MESSAGES;
    let mut client = Client::builder(&handler.config.credentials.discord, intents)
        .event_handler(handler)
        .framework(framework)
        .await?;

    {
        let mut data = client.data.write().await;
        let my_shard_manager = MyShardManager {
            shard_manager: client.shard_manager.clone(),
            runners: HashMap::new(),
        };
        data.insert::<MyShardManagerContainer>(Arc::new(Mutex::new(my_shard_manager)));
    }

    println!("Starting client...");
    client.start().await.map_err(Into::into)
}
